### PR TITLE
Updated Datadog metrics

### PR DESCRIPTION
**Pull request by bcavlin**
### :gem: Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### :scroll: Example code
```js

``` 

## :vertical_traffic_light: How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## :checkered_flag: Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] **I have added tests that prove my fix is effective or that my feature works**
- [x] **New and existing unit tests pass locally with my changes**
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by icebob** [comment]
Could you link some docs from Datadog regarding the changes?

**Comment by bcavlin** [comment]
I updated comment in the code, e.g. /* More info: https://docs.datadoghq.com/api/latest/metrics/#submit-metrics */, what documents do you refer to?

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,4 @@ benchmark/results
 .prettierrc.js
 .prettierrc
 logs/
+.history/
```
Comment (icebob) [inline .gitignore:14]: Why is it necessary?

File: src/metrics/registry.js [hunk 2]
```diff
@@ -127,7 +127,7 @@ class MetricRegistry {
 	register(opts) {
 		if (!isPlainObject(opts)) throw new Error("Wrong argument. Must be an Object.");
 
-		if (!opts.type) throw new Error("The metric 'type' property is mandatory.");
+		if (opts.type === undefined) throw new Error("The metric 'type' property is mandatory.");
```
Comment (icebob) [inline src/metrics/registry.js:130]: Why did you change it? `type == null` or `type == ""` is valid? Please revert it.

File: src/metrics/reporters/datadog.js [hunk 3]
```diff
@@ -95,18 +95,19 @@ class DatadogReporter extends BaseReporter {
...
 
-		return fetch(
-			`${this.opts.baseUrl}${this.opts.apiVersion}${this.opts.path}?api_key=${this.opts.apiKey}`,
-			{
-				method: "post",
-				body: JSON.stringify({ series }),
-				headers: {
-					"Content-Type": "application/json"
-				}
+		return fetch(`${this.opts.baseUrl}${this.opts.apiVersion}${this.opts.path}`, {
```
Comment (icebob) [inline src/metrics/reporters/datadog.js:98]: If the modified reporter only support v2, we should lock the apiVersion in the URL.

File: src/metrics/types/index.js [hunk 4]
```diff
@@ -24,9 +24,29 @@ const Types = {
  */
 function getByName(name) {
 	/* istanbul ignore next */
-	if (!name) return null;
+	if (name === undefined) return null;
```
Comment (icebob) [inline src/metrics/types/index.js:27]: Same problem, why `name == null` is valid?

File: src/metrics/types/index.js [hunk 5]
```diff
@@ -24,9 +24,29 @@ const Types = {
...
 
-	let n = Object.keys(Types).find(n => n.toLowerCase() == name.toLowerCase());
+	let converted = "Base";
+	switch (name) {
```
Comment (icebob) [inline src/metrics/types/index.js:30]: It's a common code, not just Datadog specific, don't change it, because it will affect all reporters.

File: test/unit/metrics/reporters/datadog.spec.js [hunk 6]
```diff
@@ -176,10 +176,11 @@ describe("Test Datadog Reporter class", () => {
 
 			expect(reporter.generateDatadogSeries).toBeCalledTimes(1);
 			expect(fetch).toBeCalledTimes(1);
-			expect(fetch).toBeCalledWith("https://api.datadoghq.com/api/v1/series?api_key=12345", {
+			expect(fetch).toBeCalledWith("https://api.datadoghq.com/api/v1/series", {
```
Comment (icebob) [inline test/unit/metrics/reporters/datadog.spec.js:179]: It should be `v2`, right?

File: test/unit/metrics/reporters/datadog.spec.js [hunk 7]
```diff
@@ -212,14 +213,14 @@ describe("Test Datadog Reporter class", () => {
 			});
 			reporter.init(registry);
 
-			registry.register({ name: "os.datetime.utc", type: "gauge" }).set(123456);
-			registry.register({ name: "test.info", type: "info" }).set("Test Value");
+			registry.register({ name: "os.datetime.utc", type: 3 }).set(123456);
```
Comment (icebob) [inline test/unit/metrics/reporters/datadog.spec.js:216]: The registry is common, don't change the type. You should make the Datadog specific conversion inside the Datadog reporter and not in the common classes.

**Review by icebob — CHANGES_REQUESTED** [review]

**Review by icebob — CHANGES_REQUESTED** [review]
And please revert package-lock.json

File: src/metrics/registry.js (hunk 2 above)
Comment (bcavlin) [reply src/metrics/registry.js:130]: This is because !opts.type = 0 is true as metrics changed to numbers.

File: .gitignore (hunk 1 above)
Comment (bcavlin) [reply .gitignore:14]: I have history plugin so it does not hurt to include this

File: src/metrics/reporters/datadog.js (hunk 3 above)
Comment (bcavlin) [reply src/metrics/reporters/datadog.js:98]: Not sure, up to you. Will lock it.

File: src/metrics/types/index.js (hunk 4 above)
Comment (bcavlin) [reply src/metrics/types/index.js:27]: Again, probably because of numbers? 0?

File: test/unit/metrics/reporters/datadog.spec.js (hunk 7 above)
Comment (bcavlin) [reply test/unit/metrics/reporters/datadog.spec.js:216]: But this is the test for datadog reporter, why is this common? They use numbers now, not names.

**Comment by icebob** [comment]
Something is wrong with this PR, it shows as merged but it's not, and it shows a lot of non-relevant changes as well.
